File: src/main/java/rag/retrieval/KeywordRetriever.java
package rag.retrieval;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;
import java.util.Locale;

public class KeywordRetriever implements Retriever {

    private final int topK;
    private final List<String> priorityOrder;

    public KeywordRetriever(int topK, List<String> priorityOrder) {
        this.topK = topK;
        this.priorityOrder = List.copyOf(priorityOrder);
    }

    @Override
    public List<Hit> retrieve(String question, List<String> terms, List<Document> documents) {
        if (documents == null || documents.isEmpty()) {
            return Collections.emptyList();
        }
        String normalizedQuestion = question == null ? "" : question.toLowerCase(Locale.ROOT).trim();
        List<String> normalizedTerms = terms == null ? List.of() : terms.stream()
                .filter(t -> t != null && !t.isBlank())
                .map(t -> t.toLowerCase(Locale.ROOT))
                .toList();

        List<Hit> hits = new ArrayList<>();

        for (Document doc : documents) {
            String text = doc.text() == null ? "" : doc.text();
            String lowerText = text.toLowerCase(Locale.ROOT);
            double score = 0;

            if (!normalizedQuestion.isBlank() && lowerText.contains(normalizedQuestion)) {
                score += 2;
            }

            for (String term : normalizedTerms) {
                if (lowerText.contains(term)) {
                    score += 1;
                }
            }

            if (score > 0) {
                hits.add(new Hit(doc.id(), doc.source(), doc.title(), text, score));
            }
        }

        hits.sort(Comparator
                .comparingDouble(Hit::score).reversed()
                .thenComparing(hit -> priorityIndex(hit.source()))
                .thenComparing(Hit::docId));

        if (hits.size() > topK) {
            return new ArrayList<>(hits.subList(0, topK));
        }
        return hits;
    }

    private int priorityIndex(String source) {
        int idx = priorityOrder.indexOf(source);
        return idx >= 0 ? idx : priorityOrder.size();
    }
}

File: src/main/python/rag/retrieval/keyword_retriever.py
from typing import List, Optional
from rag.retrieval.retriever import Retriever
from rag.retrieval.document import Document
from rag.retrieval.hit import Hit


class KeywordRetriever(Retriever):
    def __init__(self, top_k: int, priority_order: List[str]) -> None:
        self.top_k: int = top_k
        self.priority_order: List[str] = list(priority_order)

    def retrieve(self, question: Optional[str], terms: Optional[List[str]], documents: List[Document]) -> List[Hit]:
        if not documents:
            return []
        normalized_question: str = (question.lower().strip() if question else "")
        normalized_terms: List[str] = []
        if terms:
            normalized_terms = [t.lower() for t in terms if t and t.strip()]

        hits: List[Hit] = []

        for doc in documents:
            text: str = doc.text if doc.text else ""
            lower_text: str = text.lower()
            score: float = 0.0

            if normalized_question and normalized_question in lower_text:
                score += 2.0

            for term in normalized_terms:
                if term in lower_text:
                    score += 1.0

            if score > 0:
                hits.append(Hit(doc.id, doc.source, doc.title, text, score))

        hits.sort(key=lambda h: (
            -h.score,
            self._priority_index(h.source),
            h.doc_id
        ))

        if len(hits) > self.top_k:
            return hits[:self.top_k]
        return hits

    def _priority_index(self, source: str) -> int:
        try:
            idx: int = self.priority_order.index(source)
            return idx
        except ValueError:
            return len(self.priority_order)






